### PR TITLE
Interprocedural optimizations

### PR DESCRIPTION
**Pull request by goedderz**
### Scope & Purpose

Enable IPO/LTO in our builds.

- [X] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)
- [X] The behavior in this PR can be (and was) *manually tested*

This is expected to (but these aren't verified yet):
- Improve overall performance
- Allow moving of inline definitions in `.h` files to `.cpp` files without performance degradation, and thus
- improve compile time after many inline definitions are moved

It also allows for
- More warnings like `-Wodr` (we do currently have ODR violations!)

### Testing & Verification

This change is already covered by existing tests, such as *all of them*.

### CODE REVIEW

**Comment by goedderz** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/5476/

**Comment by goedderz** [comment]
Some changes first made here moved to https://github.com/arangodb/arangodb/pull/9666.

**Comment by ObiWahn** [comment]
The `INTERPROCEDURAL_OPTIMIZATION` property is set for every executable separately to the same value (`${IPO_ENABLED}`). What is the case against using `CMAKE_INTERPROCEDURAL_OPTIMIZATION` (https://cmake.org/cmake/help/v3.9/variable/CMAKE_INTERPROCEDURAL_OPTIMIZATION.html#variable:CMAKE_INTERPROCEDURAL_OPTIMIZATION)?

**Comment by goedderz** [comment]
Probably nothing but my ignorance. Thanks for the hint!

**Comment by goedderz** [comment]
@ObiWahn sadly, when setting `CMAKE_INTERPROCEDURAL_OPTIMIZATION`, we get dozens of warnings for 3rd party libraries:
```CMake Warning (dev) at 3rdParty/rocksdb/6.2/CMakeLists.txt:769 (add_library):
  Policy CMP0069 is not set: INTERPROCEDURAL_OPTIMIZATION is enforced when
  enabled.  Run "cmake --help-policy CMP0069" for policy details.  Use the  
  cmake_policy command to set the policy and suppress this warning.  
                                                                         
  INTERPROCEDURAL_OPTIMIZATION property will be ignored for target 'rocksdb'.        
This warning is for project developers.  Use -Wno-dev to suppress it..```

**Comment by goedderz** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6280/

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6281/

**Comment by goedderz** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6306/

File: 3rdParty/CMakeLists.txt [hunk 1]
```diff
@@ -8,6 +8,18 @@ include(ExternalProject)
 
 remove_definitions("-DUSE_ENTERPRISE=1")
 
+# Disable IPO for 3rdParty libs. This is because they currently all use old
```
Comment (ObiWahn) [inline 3rdParty/CMakeLists.txt]: nice!

File: CMakeLists.txt [hunk 2]
```diff
@@ -785,6 +785,42 @@ if (USE_FAILURE_TESTS)
   add_definitions("-DARANGODB_ENABLE_FAILURE_TESTS=1")
 endif ()
 
+################################################################################
+## INTERPROCEDURAL OPTIMIZATION (LINK TIME OPTIMIZATION)
+################################################################################
+
+set(USE_IPO AUTO CACHE STRING "Use interprocedural optimization: ON, OFF or AUTO")
```
Comment (ObiWahn) [inline CMakeLists.txt]: good!

**Comment by goedderz** [comment]
The last PR run is green, considering the new Mac run.

The package builds are also okay, but for a permission problem on Mac which is unrelated.

**Pull request merged**